### PR TITLE
Fix `assigned` and `audenceMismatch` logic, as well as other minor fixes

### PR DESCRIPTION
**Pull request by Ayesh**
- Fixes the `audienceMismatch` logic
 - Fixes the publish event not submitting `assigned` property
 - Consolidates the clock functions
 - Additional tests to ensure the publish payload is correct.
 - Minor misc fixes.

### CODE REVIEW

File: src/Context/Context.php [hunk 1]
```diff
@@ -530,9 +531,8 @@ public function flush(): void {
 
 	private function buildPublishEvent(): PublishEvent {
 		$event = new PublishEvent();
-		$event->publishedAt = $this->getTime();
 		$event->units = $this->units;
-		$event->hashed = true;
+		$event->hashed = false;
```
Comment (marcio-absmartly) [inline src/Context/Context.php:535]: This cannot be correct, right? All SDKs always publish with `hashed = true`.

File: src/PublishEvent.php [hunk 2]
```diff
@@ -13,9 +15,7 @@ class PublishEvent implements JsonSerializable {
 	public array $attributes;
 
 	public function jsonSerialize(): object {
-		$object = (object) [
-			'publishedAt' => $this->publishedAt,
-		];
+		$object = new stdClass();
 
 		foreach (['hashed', 'exposures', 'attributes'] as $key) {
```
Comment (marcio-absmartly) [inline src/PublishEvent.php:20]: It seems like exposures are already taken care of in the loop with the goals below (only visible here on github if you expand the code below), right?

**Review by marcio-absmartly — CHANGES_REQUESTED** [review]
I think the `assigned` part is fixed.
It also seems that the first for the bad format on the hashes was to send `hashed = false`, but this is not acceptable.
The tests are still not validating the whole payload, only the exposures part. We also need to ensure the hashed units and all the other bits of the payload are matching 100%.

File: src/PublishEvent.php (hunk 2 above)
Comment (Ayesh) [reply src/PublishEvent.php:20]: You are right that was mixed up there. Fix coming up.

**Comment by Ayesh** [comment]
Hi @marcio-absmartly - thank you for your patience and the thorough review with this.
You are right there was no parity with the other SDKs that hashed the data by themselves and marked the `hashed` value as `true`. I have applied a similar logic to the `PublishEvent` class. 

 - You will notice the URL-safe base64 encoding method was copied from `VariantAssigner`. Despite DRY, I thought to copy it over because that function is relatively small and contained, and to not trigger an autoload for the `VariantAssigner`. There are some discussions in PHP to add a URL-safe base64 encoding to PHP core as well, and that was another motivator too. 

 - I also included some changes to the `HTTPClient` to show the error message contents, which was not working properly before. While this has no direct impact to the bug we are trying to fix now, this is in preparation for additional HTTPClient tests to add error message assertions as well as error code assertions.

**Comment by marcio-absmartly** [comment]
@Ayesh seems like your fixes are working, however I cannot test it end to end, because now it's failing on the goals.
I found a few more critical issues before we can push this to customers:
- `publishedAt` in the publish payload is not initialized, this is crucial.
- All timings seem to be truncated to seconds, we need millisecond precision here, it is also crucial.
You cannot see this in the Events page of the demo cluster, because we do some calculation after ingestion, but if you just print the payload, you will see it.

Once again, the Java tests are ensuring the full payload, not just exposures/goals are sent correctly, so this should be tested for as well in PHP.

File: src/Context/Context.php [hunk 3]
```diff
@@ -74,7 +74,7 @@ public function isClosed(): bool {
 	}
 
 	public static function getTime(): int {
-		return (int) microtime(true) * 1000;
+		return (int) (microtime(true) * 1000);
```
Comment (marcio-absmartly) [inline src/Context/Context.php:77]: Are those ints 64-bit? :)

Comment (Ayesh) [reply src/Context/Context.php:77]: Yes, as long as it runs on 64 bit :)

There are extensions that support distinct bit widths, but PHP core only has this. We could check for 32 bit processors and perhaps set this value as a string, but I'm not sure if the API accepts it? I'll happily work that out if that's the case.

Comment (Ayesh) [reply src/Context/Context.php:77]: Also, we can define a clause in the `composer.json` file to require a 64-bit processor. Composer will refuse to install the SDK on 32 bit processors.

Comment (Ayesh) [reply src/Context/Context.php:77]: #2 proposes this.

**Comment by marcio-absmartly** [comment]
Hi @Ayesh,

The only remaining thing seems to be the attributes being stored internally as a dict, as I don't see all the sets in the payload.

**Review by marcio-absmartly — APPROVED** [review]

**Pull request merged**